File: packages/python/src/Seguranca/Ataques/AtaqueRotacional.py
"""Ataque rotacional/slide: se a cifra for simétrica a deslocamentos circulares
de key/iv, então `keystream(rot(key), rot(iv))` seria uma rotação de
`keystream(key, iv)` - uma estrutura explorável. Testa todas as rotações de
byte e também coincidência direta do keystream.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import random_bytes


def rotacionar(buf: bytes, k: int) -> bytes:
    n = len(buf)
    return bytes(buf[(i + k) % n] for i in range(n))


class AtaqueRotacional:
    def __init__(self, tamanho: int = 64):
        self.tamanho = tamanho

    def nome(self) -> str:
        return "Rotacional/slide (simetria por rotação)"

    def executar(self, alvo) -> ResultadoAtaque:
        chave = alvo.chave_de_teste().encode()
        iv = random_bytes(alvo.tamanho_iv())
        ks1 = alvo.gerar_keystream_bruto(chave, iv, "enc", self.tamanho)

        coincidencias = []
        for k in range(1, len(chave)):
            ks2 = alvo.gerar_keystream_bruto(
                rotacionar(chave, k),
                rotacionar(iv, k),
                "enc",
                self.tamanho,
            )

            if ks2 == ks1:
                coincidencias.append(f"rotação {k}: keystream idêntico")
                continue
            # metade inicial de ks1 rotacionada deve bater com ks2 se houver simetria
            alvo_rot = rotacionar(ks1[:32], k)
            if ks2[:32] == alvo_rot:
                coincidencias.append(f"rotação {k}: keystream rotacionado")

        vulneravel = len(coincidencias) > 0

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "critica" if vulneravel else "info",
            (
                f"Simetria rotacional encontrada: {'; '.join(coincidencias[:5])}"
                if vulneravel
                else f"Nenhuma das {len(chave) - 1} rotações de byte reproduziu o keystream"
            ),
            {"coincidencias": coincidencias},
        )

File: packages/rust/src/seguranca/ataques/ataque_rotacional.rs
//! Ataque rotacional/slide: se a cifra for simétrica a deslocamentos circulares
//! de key/iv, então `keystream(rot(key), rot(iv))` seria uma rotação de
//! `keystream(key, iv)` - uma estrutura explorável. Testa todas as rotações de
//! byte e também coincidência direta do keystream.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueRotacional {
    pub tamanho: usize,
}

impl AtaqueRotacional {
    pub fn new() -> Self {
        AtaqueRotacional { tamanho: 64 }
    }
}

impl Default for AtaqueRotacional {
    fn default() -> Self {
        Self::new()
    }
}

impl AtaqueRotacional {
    fn rotacionar(buf: &[u8], k: usize) -> Vec<u8> {
        let n = buf.len();
        (0..n).map(|i| buf[(i + k) % n]).collect()
    }
}

impl Ataque for AtaqueRotacional {
    fn nome(&self) -> String {
        "Rotacional/slide (simetria por rotação)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();
        let chave = chave.as_bytes();
        let iv = random_bytes(alvo.tamanho_iv());
        let ks1 = alvo.gerar_keystream_bruto(chave, &iv, "enc", self.tamanho);

        let mut coincidencias: Vec<String> = Vec::new();
        for k in 1..chave.len() {
            let ks2 = alvo.gerar_keystream_bruto(
                &Self::rotacionar(chave, k),
                &Self::rotacionar(&iv, k),
                "enc",
                self.tamanho,
            );

            if ks2 == ks1 {
                coincidencias.push(format!("rotação {}: keystream idêntico", k));
                continue;
            }
            // metade inicial de ks1 rotacionada deve bater com ks2 se houver simetria
            let alvo_rot = Self::rotacionar(&ks1[..32], k);
            if ks2[..32] == alvo_rot[..] {
                coincidencias.push(format!("rotação {}: keystream rotacionado", k));
            }
        }

        let vulneravel = !coincidencias.is_empty();

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Critica
            } else {
                Severidade::Info
            },
            &if vulneravel {
                let amostra: Vec<String> = coincidencias.iter().take(5).cloned().collect();
                format!("Simetria rotacional encontrada: {}", amostra.join("; "))
            } else {
                format!(
                    "Nenhuma das {} rotações de byte reproduziu o keystream",
                    chave.len() - 1
                )
            },
        ))
    }
}
